framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java

```diff
@@ -1323,6 +1327,11 @@ protected void addJoins(StringBuilder str, Collection<JoinBuilder<SearchCriteria
     public Pair<List<T>, Integer> searchAndDistinctCount(final SearchCriteria<T> sc, final Filter filter) {
         List<T> objects = search(sc, filter, null, false);
         Integer count = getDistinctCount(sc);
+        // Count cannot be 0 if there is at least a result in the list, see CLOUDSTACK-10320
+        if (count == 0 && !objects.isEmpty()) {
+            // Cannot assume if it's more than one since the count is distinct vs search
+            count = 1;
```

Comment:
why not put count = objects.size() ?

Reply:
Because the search is mostly done against views which don't return distinct rows due to the *not so smart* queries fetching the details rows. Look at the APIHelper with iterate on duplicate row entries, but only append the details.

Reply:
Moreover, the search might return a page size list, while the count should be global: listVirtualMachines returns a count of 234 and you have 50 entries in the search based on the `page` and `pageSize`.